From edbcb3864db237dcee9f2bba376652158e75d7be Mon Sep 17 00:00:00 2001
From: Peter Dreuw <peter.dreuw@credativ.de>
Date: Thu, 14 Sep 2017 14:14:43 +0200
Subject: [PATCH 11/16] Subject: grant_table: fix GNTTABOP_cache_flush handling

Don't fall over a NULL grant_table pointer when the owner of the domain
is a system domain (DOMID_{XEN,IO} etc).

This is XSA-232.

Reported-by: Matthew Daley <mattd@bugfuzz.com>
Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Jan Beulich <jbeulich@suse.com>
---
 xen/common/grant_table.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
index 60616c83f..2b9a0e06e 100644
--- a/xen/common/grant_table.c
+++ b/xen/common/grant_table.c
@@ -2637,7 +2637,7 @@ static int __gnttab_cache_flush(gnttab_cache_flush_t *cflush,
     if ( (cflush->offset >= PAGE_SIZE) ||
          (cflush->length > PAGE_SIZE) ||
          (cflush->offset + cflush->length > PAGE_SIZE) )
-        return -EINVAL;
+        return -EINVAL; 
 
     if ( cflush->length == 0 || cflush->op == 0 )
         return 0;
@@ -2660,7 +2660,7 @@ static int __gnttab_cache_flush(gnttab_cache_flush_t *cflush,
 
     page = mfn_to_page(mfn);
     owner = page_get_owner_and_reference(page);
-    if ( !owner )
+    if ( !owner || !owner->grant_table )
     {
         rcu_unlock_domain(d);
         return -EPERM;
-- 
2.14.1

